### ISSUE TITLE
Add chunk data translator

### ISSUE THREAD

**Issue by MrPowerGamerBR** [PocketDreams/GlowstonePocket#1]
Trying to send a proper PE chunk packet is hard, so the project is kinda on pause until I can properly send a chunk packet to a Pocket client.

**Comment by MrPowerGamerBR** [comment]
I will probably copy the chunk generator code from Nukkit and try to send it to see if it will work.

The problem is that I don't know what are the required packets to be sent before the chunk data packet.

**Comment by caspervanneck** [comment]
Cool @MrPowerGamerBR  :3

**Comment by caspervanneck** [comment]
@MrPowerGamerBR  chunks working

**Comment by DaMatrix** [comment]
WHAT DID HE JUST SAY???????

**Comment by DaMatrix** [comment]
He said that the project is on hold until he can figure out how to send a chunk packet to the client.

**Comment by MrPowerGamerBR** [comment]
I didn't implement chunks yet, it is just too damn difficult to be
implemented.

Em 24 de fev de 2017 08:59, "DaPorkchop" <notifications@github.com>
escreveu:

**Comment by caspervanneck** [comment]
sorry ): @DaMatrix  @MrPowerGamerBR

**Comment by MrPowerGamerBR** [comment]
And, as I said, sending a chunk to a MCPE/MCW10E client is difficult
because there are some packets that must be sent before. (And the chunk
payload data is kinda strange)

That's why this project is in hold (and because my main focus is Nukkit
(SequinLand) and MC 1.11.2 there is way more advanced than this)

I will probably try to fix the chunk data again today (If I can get a empty
chunk data to be successfully received by the client, then that's already a
HUGE achievement), but that's too difficult.

Em 24 de fev de 2017 09:04, "caspervanneck" <notifications@github.com>
escreveu:

**Comment by MrPowerGamerBR** [comment]
Sucessfully sent a chunk packet to the client 🎉

**Comment by MrPowerGamerBR** [comment]
![http://i.imgur.com/kMaAzwl.png](http://i.imgur.com/kMaAzwl.png)

**Comment by MrPowerGamerBR** [comment]
However I still need to fix the player's spawn position

**Comment by MrPowerGamerBR** [comment]
The client is still... "broken" when receiving the chunk packet, it starts flickering a lot **but** you can see that yes, the chunk packet is successfully received by the client

**Comment by MrPowerGamerBR** [comment]
Fixed the client brokeness when receiving the packet, thanks Polo1K (and MiNET's wiki page)

![http://i.imgur.com/7c72KVX.png](http://i.imgur.com/7c72KVX.png)

**Comment by MrPowerGamerBR** [comment]
![http://i.imgur.com/fVFWteq.png](http://i.imgur.com/fVFWteq.png)

I still need to compress those chunks

**Comment by robotman3000** [comment]
@MrPowerGamerBR I have code for handling batch packets. This is a [gist](https://gist.github.com/robotman3000/a16a9e33213e759b6ec3f6d7cc745b5c) for my RakNet Utility class. It uses classes and code from nukkit but it works.

**Comment by robotman3000** [comment]
lol, then I go and look at your code and find that you are 10 steps ahead of me

**Comment by MrPowerGamerBR** [comment]
@robotman3000 I already know how to handle batch packets, I just didn't write the correct code to create batch packets.

Also, if you are curious, the project is kinda in a "hey it works but it still needs more works", here's what you can do:

* Login into the server (need online-mode: false, probably won't even implement online mode support)
* Walk around into the server
* Load chunks
* View other PC/PE players
* View spawned entities
* Talk into the chat (can't execute commands without using "!" yet because I didn't implement that)

![http://i.imgur.com/9XY8EIH.png](http://i.imgur.com/9XY8EIH.png)

**Comment by MrPowerGamerBR** [comment]
And yes, I forgot to close this issue, chunks are already supported.

**Comment by MrPowerGamerBR** [comment]
And instead of doing a system like DragonProxy/Dragonet, I'm not doing PC -> PE (and soon no PE -> PC translators) translators because that would be too hacky and bad, you can't have a 100% functional server relying on PC packets due to the differences between the Notchian protocol and the Pocket protocol

**Comment by robotman3000** [comment]
@MrPowerGamerBR So how do you plan to do it? From looking at you code I can see bits and pieces of the design of the translator system from DragonProxy. I am interested in contributing to this project

**Comment by MrPowerGamerBR** [comment]
@robotman3000 instead of relying on PC packets, I send the PE packets directly from the GlowPlayer (and other classes) class.

This has a way better flexibility due to the differences between the notchian and the pocket protocol.

Here's an example:
```
    public void join(GlowSession session, PlayerReader reader) {
        // send join game
        // in future, handle hardcore, difficulty, and level type
        String type = world.getWorldType().getName().toLowerCase();
        int gameMode = getGameMode().getValue();
        if (server.isHardcore()) {
            gameMode |= 0x8;
        }
        if (this.isPocketProtocol()) {
            this.getPocketSession().sendPocket(new StartGamePacket((long) SELF_ID, (long) SELF_ID, (float) location.getX(), (float) (location.getY() + 1.62F), (float) location.getZ(), 1337, (byte) world.getEnvironment().getId(), 1, 1, world.getDifficulty().getValue(), world.getSpawnLocation().getBlockX(), world.getSpawnLocation().getBlockY(), world.getSpawnLocation().getBlockZ(), true, -1, false, 0, 0, true, false, "", "Shantae is cute").andEncode());
        } else {
            session.send(new JoinGameMessage(SELF_ID, gameMode, world.getEnvironment().getId(), world.getDifficulty().getValue(), session.getServer().getMaxPlayers(), type, world.getGameRuleMap().getBoolean("reducedDebugInfo")));
        }
        setGameModeDefaults();

        // send server brand and supported plugin channels
        session.send(PluginMessage.fromString("MC|Brand", server.getName()));
        sendSupportedChannels();
        joinTime = System.currentTimeMillis();
        reader.readData(this);
        reader.close();

        // Add player to list of online players
        getServer().setPlayerOnline(this, true);

        // save data back out
        saveData();

        streamBlocks(); // stream the initial set of blocks
        setCompassTarget(world.getSpawnLocation()); // set our compass target
        sendTime();
        sendWeather();
        sendRainDensity();
        sendSkyDarkness();
        sendAbilities();

        scoreboard = server.getScoreboardManager().getMainScoreboard();
        scoreboard.subscribe(this);

        invMonitor = new InventoryMonitor(getOpenInventory());
        updateInventory(); // send inventory contents

        // send initial location
        if (this.isPocketProtocol()) {
            this.getPocketSession().sendPocket(new PlayStatusPacket(PlayStatusPacket.SPAWNED).andEncode());
        } else {
            session.send(new PositionRotationMessage(location));
        }
        
        if (!server.getResourcePackURL().isEmpty()) {
            setResourcePack(server.getResourcePackURL(), server.getResourcePackHash());
        }
    }
```

If doing a proxy-like system, a wall would be hit eventually. (Notchian protocol uses relative movements for entities while Pocket uses absolutes), proxy also has the server desync issue, where the client is somewhere but in the server it is in another place. (DragonProxy had this issue)

**Comment by MrPowerGamerBR** [comment]
@robotman3000 the only thing that still uses an translator system is the PE -> PC packets, because trying to just direct handle those will be difficult (and probably not necessary... yet)

**Comment by MrPowerGamerBR** [comment]
@robotman3000 another example
```
        // add entities
        knownChunks.parallelStream().forEach(key -> {
            GlowChunk chunk = world.getChunkAt(key.getX(), key.getZ());
            chunk.getRawEntities().stream()
                    .filter(entity -> this != entity)
                    .filter(this::isWithinDistance)
                    .filter(entity -> !entity.isDead())
                    .filter(entity -> !knownEntities.contains(entity))
                    .filter(entity -> !hiddenEntities.contains(entity.getUniqueId()))
                    .forEach((entity) -> {
                        knownEntities.add(entity);
                        if (this.isPocketProtocol()) {
                            entity.createSpawnMessageForPocket().forEach((packet) -> this.getPocketSession().sendPocket(packet));
                        } else {
                            entity.createSpawnMessage().forEach(session::send);
                        }
                    });
        });
```

**Comment by robotman3000** [comment]
@MrPowerGamerBR So in this fork is the design to replace the pc system or exist at the same time?

**Comment by robotman3000** [comment]
Exist at the same time right?

**Comment by MrPowerGamerBR** [comment]
@robotman3000 yeah, both at the same time (that's why there is a this.isPocketProtocol() check)

Also, the project *works*, but requires a bit of tweaking to work, you need to disable online-mode in glowstone.yml and decrease view-distance to something lower (2)

**Comment by robotman3000** [comment]
what classes are these examples in?

**Comment by MrPowerGamerBR** [comment]
@robotman3000 both are inside the net.glowstone.entity.GlowPlayer

**Comment by robotman3000** [comment]
@MrPowerGamerBR Is the GlowPlayer class the place where most of the packet sending occurs?

**Comment by MrPowerGamerBR** [comment]
@robotman3000 yes, there are some classes that also sends packets (GlowWorld) and some packets are created in another classes (GlowCreature), but most of them is on the GlowPlayer class.

PE -> PC packet translator happens in SequinLand's translator classes.

**Comment by MrPowerGamerBR** [comment]
@robotman3000 but most of the code is like that:

if is pocket edition:
    send pocket edition packet
else:
    send original packet

**Comment by MrPowerGamerBR** [comment]
@robotman3000 packet SENDING (where packets are... sent, but not created!) is handled in PocketSession

**Comment by robotman3000** [comment]
@MrPowerGamerBR What I'm thinking is that it might be a better design to make GlowPlayer abstract or an interface and have two classes that extend it. I.E. `GlowPCPlayer` and `GlowPEPlayer` such that when you call `setCompassTarget`, for example, the implementation of `setCompassTarget` in `GlowPCPlayer` would generate the PC message that sets the compass target and `setCompassTarget` in `GlowPEPlayer` would generate the PE version of that message.

The idea is that when an "event" occurs on the server where packets would be sent, the method for that "event" (Like `setCompassTarget`) would be called and the implementation of GlowPlayer wold handle the actual message generation

**Comment by robotman3000** [comment]
@MrPowerGamerBR This way we avoid checking if we are PE or PC and just let the OO nature of java do the "thinking" for us

**Comment by MrPowerGamerBR** [comment]
@robotman3000 I was going to do that, but Glowstone's source code has a lot of references to "GlowPlayer.class", so doing that without heavily modifying the source code is a bit difficult.

**Comment by robotman3000** [comment]
In what ways is `GlowPlayer.class` used?

**Comment by MrPowerGamerBR** [comment]
@robotman3000
![http://i.imgur.com/bJILgjV.png](http://i.imgur.com/bJILgjV.png)

The most "wow" one is the saving one.

**Comment by MrPowerGamerBR** [comment]
@robotman3000 but well, if you can do that, go ahead, it would be better than I'm doing now. 😉 

(Also, if you need a more "instant" response, SequinLand has a discord here: https://discord.gg/EZArH2G )

**Comment by robotman3000** [comment]
Well, my time is very limited but perhaps I can try to make this idea work.

**Comment by robotman3000** [comment]
How would I start the "project"? Should I fork this repo and work there?

**Comment by MrPowerGamerBR** [comment]
@robotman3000 if you want repo access, I can give you that

**Comment by robotman3000** [comment]
That wold be nice. Would you want me to make a new branch or push to the master branch?

**Comment by MrPowerGamerBR** [comment]
@robotman3000 if it is doesn't break (so, if it still works when compiling and connecting) then push to the master branch, if not, push to a separate branch

**Comment by robotman3000** [comment]
Ok then, thanks. It might be a few days before I can work on this, though.

**Comment by MrPowerGamerBR** [comment]
@robotman3000 okay
